pgrx/src/spi/tuple.rs

```diff
@@ -0,0 +1,529 @@
+use std::ffi::{CStr, CString};
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::ops::Index;
+use std::ptr::NonNull;
+
+use crate::memcxt::PgMemoryContexts;
+use crate::pg_sys::{self, PgOid};
+use crate::prelude::*;
+use crate::spi::Result as SpiResult;
+
+use super::{Error, SpiErrorCodes, SpiOkCodes};
+use pgrx_pg_sys::panic::ErrorReportable;
+
+#[derive(Debug)]
+pub struct SpiTupleTable<'conn> {
+    #[allow(dead_code)]
+    pub(super) status_code: SpiOkCodes,
+    pub(super) table: Option<&'conn mut pg_sys::SPITupleTable>,
+    pub(super) size: usize,
+    pub(super) current: isize,
+}
+
+impl<'conn> SpiTupleTable<'conn> {
+    /// `SpiTupleTable`s are positioned before the start, for iteration purposes.
+    ///
+    /// This method moves the position to the first row.  If there are no rows, this
+    /// method will silently return.
+    pub fn first(mut self) -> Self {
+        self.current = 0;
+        self
+    }
+
+    /// Restore the state of iteration back to before the start.
+    ///
+    /// This is useful to iterate the table multiple times
+    pub fn rewind(mut self) -> Self {
+        self.current = -1;
+        self
+    }
```

Comment:
```suggestion
impl<'conn> SpiTupleTable<'conn> {
    /// `SpiTupleTable`s are positioned before the start, for iteration purposes.
    ///
    /// This method moves the position to the first row.  If there are no rows, this
    /// method will silently return.
    pub fn first(mut self) -> Self {
        self.current = 0;
        self
    }

    /// Restore the state of iteration back to before the start.
    ///
    /// This is useful to iterate the table multiple times
    pub fn rewind(mut self) -> Self {
        self.current = -1;
        self
    }
```

These functions should probably instead use `&mut self`. They could still return `&mut Self` if we really like that style of API.

Reply:
sure.